src/libAtomVM/opcodesswitch.h

```diff
@@ -3243,10 +3243,6 @@ HOT_FUNC int scheduler_entry_point(GlobalContext *glb)
                 DECODE_DEST_REGISTER(dreg, dreg_type, code, i, next_off);
 
 #ifdef IMPL_EXECUTE_LOOP
-                // MEMORY_NO_SHRINK because put_list is classified as put in beam_ssa_codegen.erl
-                if (UNLIKELY(memory_ensure_free_opt(ctx, CONS_SIZE, MEMORY_NO_SHRINK) != MEMORY_GC_OK)) {
-                    RAISE_ERROR(OUT_OF_MEMORY_ATOM);
-                }
```

Comment:
Why is this safe to remove?  Was it added in #765 and being retracted?

Reply:
That's the point of this PR.

I got the fix wrong in #765. I thought there was no heap allocation opcode added by the compiler before the opcodes it classifies as put, the same way there are none before the opcodes it classifies as gc, and that was the cause of the crash.

```
{function, disp, 2, 7}.
  {label,6}.
    {line,[{location,"../tests/erlang_tests/test_put_without_heap_allocation.erl",
                     38}]}.
    {func_info,{atom,test_put_without_heap_allocation},{atom,disp},2}.
  {label,7}.
    {'%',{var_info,{x,1},[{type,{t_integer,{0,255}}}]}}.
    {allocate,2,2}.
    {move,{x,1},{y,0}}.
    {move,{x,0},{y,1}}.
    {move,{x,1},{x,0}}.
    {line,[{location,"../tests/erlang_tests/test_put_without_heap_allocation.erl",
                     39}]}.
    {call_ext,1,{extfunc,erlang,integer_to_binary,1}}.
    {line,[{location,"../tests/erlang_tests/test_put_without_heap_allocation.erl",
                     41}]}.
    {gc_bif,byte_size,{f,0},1,[{x,0}],{x,1}}.
    {bs_add,{f,0},[{x,1},{integer,5},1],{x,1}}.
    {bs_init2,{f,0},{x,1},7,2,{field_flags,[]},{x,1}}.
    {bs_put_string,5,{string,<<"Test ">>}}.
    {bs_put_binary,{f,0},{atom,all},8,{field_flags,[unsigned,big]},{x,0}}.
    {put_tuple2,{x,0},{list,[{atom,text},{integer,10},{y,0},{x,1}]}}.
    {put_list,{x,0},nil,{x,1}}.
    {move,{y,1},{x,0}}.
    {line,[{location,"../tests/erlang_tests/test_put_without_heap_allocation.erl",
                     43}]}.
    send.
    {deallocate,2}.
    return.
```

Indeed there are no `test_heap` or `allocate` here between `gc_bif` that can shrink the heap and `put_tuple2`. However, it sounded weird, plus Erlang/OTP doesn't have allocation test in [`put_tuple2`](https://github.com/erlang/otp/blob/7f9f8f7128c1016000194635e300f5e07c9a9b24/erts/emulator/beam/emu/instrs.tab#L726).

But in the code above, there is a little **7** as a parameter of `bs_init2` that we ignored. I figured this out by making some code that needed much more than 7 terms (and that didn't crash Erlang/OTP) as explained in [a comment](https://github.com/atomvm/AtomVM/pull/765#issuecomment-1688919587) in #765.

Reply:
Makes sense